### PR TITLE
Update snapshot CRD to v1beta

### PR DESCRIPTION
**Pull request by xing-yang**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR updates dependencies to VolumeSnapshot CRD v1beta1.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Note: Dependency files from external-snapshotter are manually copied at this point.
Snapshot beta APIs are not finalized yet.

**Does this PR introduce a user-facing change?**:
```release-note
Updates VolumeSnapshot CRD to v1beta1
```

### CODE REVIEW

File: vendor/github.com/kubernetes-csi/external-snapshotter/pkg/apis/volumesnapshot/v1beta1/types.go [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+	// driver is the name of the CSI driver used to create the physical snapshot on
+	// the underlying storage system.
+	// This MUST be the same as the name returned by the CSI GetPluginName() call for
+	// that driver.
+	// Required.
+	Driver string `json:"driver" protobuf:"bytes,3,opt,name=driver"`
+
+	// name of the SnapshotClass to which this snapshot belongs.
+	// +optional
+	SnapshotClassName *string `json:"snapshotClassName,omitempty" protobuf:"bytes,4,opt,name=snapshotClassName"`
```
Comment (xing-yang) [inline vendor/github.com/kubernetes-csi/external-snapshotter/pkg/apis/volumesnapshot/v1beta1/types.go]: We should call it VolumeSnapshotClassName to be consistent with the same thing in VolumeSnapshotSpec.  VolumeSnapshotClassName was also the original name in the content spec.

**Comment by xing-yang** [comment]
Hi @msau42, this PR uses the latest external-snapshotter master code.  It is ready for review.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -808,36 +808,31 @@ func (p *csiProvisioner) getSnapshotSource(options controller.ProvisionOptions)
 	}
 	klog.V(5).Infof("VolumeSnapshot %+v", snapshotObj)
 
-	snapContentObj, err := p.snapshotClient.VolumesnapshotV1alpha1().VolumeSnapshotContents().Get(snapshotObj.Spec.SnapshotContentName, metav1.GetOptions{})
-	if err != nil {
-		klog.Warningf("error getting snapshotcontent %s for snapshot %s/%s from api server: %s", snapshotObj.Spec.SnapshotContentName, snapshotObj.Namespace, snapshotObj.Name, err)
-		return nil, fmt.Errorf("snapshot in dataSource not bound or invalid")
+	if snapshotObj.Status.BoundVolumeSnapshotContentName == nil {
+		return nil, fmt.Errorf("snapshot %s does not have a content", options.PVC.Spec.DataSource.Name)
```
Comment (msau42) [inline pkg/controller/controller.go]: how about "snapshot %s not bound"?

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -795,11 +795,11 @@ func (p *csiProvisioner) getPVCSource(options controller.ProvisionOptions) (*csi
 // getSnapshotSource verifies DataSource.Kind of type VolumeSnapshot, making sure that the requested Snapshot is available/ready
 // returns the VolumeContentSource for the requested snapshot
 func (p *csiProvisioner) getSnapshotSource(options controller.ProvisionOptions) (*csi.VolumeContentSource, error) {
-	snapshotObj, err := p.snapshotClient.VolumesnapshotV1alpha1().VolumeSnapshots(options.PVC.Namespace).Get(options.PVC.Spec.DataSource.Name, metav1.GetOptions{})
+	snapshotObj, err := p.snapshotClient.SnapshotV1beta1().VolumeSnapshots(options.PVC.Namespace).Get(options.PVC.Spec.DataSource.Name, metav1.GetOptions{})
 	if err != nil {
 		return nil, fmt.Errorf("error getting snapshot %s from api server: %v", options.PVC.Spec.DataSource.Name, err)
 	}
-	if snapshotObj.Status.ReadyToUse == false {
+	if snapshotObj.Status.ReadyToUse == nil || *snapshotObj.Status.ReadyToUse == false {
```
Comment (msau42) [inline pkg/controller/controller.go]: Can we move this check below the status.Bound check?  I think it makes sense in terms of ordering to:
1. Check for deletion
2. Check for bound
3. Check if ready

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -808,36 +808,31 @@ func (p *csiProvisioner) getSnapshotSource(options controller.ProvisionOptions)
...
-		klog.Warningf("snapshotcontent %s for snapshot %s/%s is not bound", snapshotObj.Spec.SnapshotContentName, snapshotObj.Namespace, snapshotObj.Name)
-		return nil, fmt.Errorf("snapshot in dataSource not bound or invalid")
-	}
+	snapContentObj, err := p.snapshotClient.SnapshotV1beta1().VolumeSnapshotContents().Get(*snapshotObj.Status.BoundVolumeSnapshotContentName, metav1.GetOptions{})
 
-	if snapContentObj.Spec.VolumeSnapshotRef.UID != snapshotObj.UID || snapContentObj.Spec.VolumeSnapshotRef.Namespace != snapshotObj.Namespace || snapContentObj.Spec.VolumeSnapshotRef.Name != snapshotObj.Name {
-		klog.Warningf("snapshotcontent %s for snapshot %s/%s is bound to a different snapshot", snapshotObj.Spec.SnapshotContentName, snapshotObj.Namespace, snapshotObj.Name)
+	if err != nil {
+		klog.Warningf("error getting snapshotcontent %s for snapshot %s/%s from api server: %s", *snapshotObj.Status.BoundVolumeSnapshotContentName, snapshotObj.Namespace, snapshotObj.Name, err)
 		return nil, fmt.Errorf("snapshot in dataSource not bound or invalid")
```
Comment (msau42) [inline pkg/controller/controller.go]: maybe change all these user-visible error messages to the same as above "snapshot %s not bound" (and make it a constant)

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -808,36 +808,31 @@ func (p *csiProvisioner) getSnapshotSource(options controller.ProvisionOptions)
...
+	if snapContentObj.Spec.Driver != options.StorageClass.Provisioner {
+		klog.Warningf("snapshotcontent %s for snapshot %s/%s is handled by a different CSI driver than requested by StorageClass %s", *snapshotObj.Status.BoundVolumeSnapshotContentName, snapshotObj.Namespace, snapshotObj.Name, options.StorageClass.Name)
 		return nil, fmt.Errorf("snapshot in dataSource not bound or invalid")
 	}
 
 	klog.V(5).Infof("VolumeSnapshotContent %+v", snapContentObj)
 	snapshotSource := csi.VolumeContentSource_Snapshot{
 		Snapshot: &csi.VolumeContentSource_SnapshotSource{
-			SnapshotId: snapContentObj.Spec.VolumeSnapshotSource.CSI.SnapshotHandle,
+			SnapshotId: *snapContentObj.Status.SnapshotHandle,
```
Comment (msau42) [inline pkg/controller/controller.go]: Could this be nil?

Comment (xing-yang) [reply pkg/controller/controller.go]: Yes.  Added a check.

File: pkg/controller/controller.go (hunk 4 above)
Comment (xing-yang) [reply pkg/controller/controller.go]: Sure

File: pkg/controller/controller.go (hunk 3 above)
Comment (xing-yang) [reply pkg/controller/controller.go]: Sure

File: pkg/controller/controller.go (hunk 2 above)
Comment (xing-yang) [reply pkg/controller/controller.go]: Sure

File: pkg/controller/controller.go (hunk 5 above)
Comment (msau42) [reply pkg/controller/controller.go]: Can we also add unit tests for these various new nil checks?  It may be simpler to restructure the test to pass in the exact content object to use, or use a builder pattern

Comment (xing-yang) [reply pkg/controller/controller.go]: sure.

**Comment by xing-yang** [comment]
@msau42 unit tests added. PTAL.  Thanks.
